### PR TITLE
Some additions to the Path Tracing & Lighting render controls

### PR DESCRIPTION
**Pull request by Mango-3**
- Add `Low Light Threshold`, enable `Direct Lighting` & enable `Image-Based Lighting` 
controls to the `Path Tracing` render controls
- Add  `Light Sampler` selection menu and lights `Importance Sampling` control to the `Lighting` render controls
- Add slider controls to numeric render parameter. 
The slider parameter ranges (soft-limits) are chosen to be in a suitable range for 99% of use cases but can be easily overridden in the numeric input fields up to the hard limits of each parameter.

### CODE REVIEW

File: scripts/appleseedMaya/renderGlobals.py [hunk 1]
```diff
@@ -391,10 +391,11 @@ def create(self):
...
             with pm.columnLayout("appleseedColumnLayout", adjustableColumn=True, width=columnWidth):
                 with pm.frameLayout(label="Sampling", collapsable=True, collapse=False):
-                    with pm.columnLayout("appleseedColumnLayout", adjustableColumn=True, width=columnWidth):
+                    with pm.columnLayout("appleseedColumnLayout", adjustableColumn=False, width=columnWidth):
                         self._addControl(
-                            ui=pm.intFieldGrp(
-                                label="Render Passes", numberOfFields=1),
+                            ui=pm.intSliderGrp(
+                                label="Render Passes", field=True, value=1, cw=(3,160), minValue=1, 
+                                fieldMinValue=1, maxValue=100, fieldMaxValue=1000000000),
```
Comment (dictoon) [inline scripts/appleseedMaya/renderGlobals.py]: I suggest using the exact same hard limits as in appleseed.studio, that would be 1000000 (one million) for the number of passes.

File: scripts/appleseedMaya/renderGlobals.py [hunk 2]
```diff
@@ -405,20 +406,24 @@ def create(self):
...
-                                label="Min Samples", numberOfFields=1, enable=adaptiveSampling),
+                            ui=pm.intSliderGrp(
+                                label="Min Samples", field=True, value=16, cw=(3,160), minValue=0, 
+                                fieldMinValue=0, maxValue=256, fieldMaxValue=1000000, enable=adaptiveSampling),
                             attrName="minPixelSamples")
                         self._addControl(
-                            ui=pm.intFieldGrp(
-                                label="Max Samples", numberOfFields=1),
+                            ui=pm.intSliderGrp(
+                                label="Max Samples", field=True, value=256, cw=(3,160), minValue=16,
```
Comment (dictoon) [inline scripts/appleseedMaya/renderGlobals.py:415]: Hard min value for Max Samples should be 0 as in appleseed.studio. Whether 0 is a reasonable hard min value or not is debatable, but at least that would be consistent with .studio.

File: scripts/appleseedMaya/renderGlobals.py [hunk 3]
```diff
@@ -405,20 +406,24 @@ def create(self):
...
-                                label="Max Samples", numberOfFields=1),
+                            ui=pm.intSliderGrp(
+                                label="Max Samples", field=True, value=256, cw=(3,160), minValue=16, 
+                                fieldMinValue=0, maxValue=1024, fieldMaxValue=1000000),
                             attrName="samples")
                         self._addControl(
-                            ui=pm.intFieldGrp(
-                                label="Batch Sample Size", numberOfFields=1, enable=adaptiveSampling),
+                            ui=pm.intSliderGrp(
+                                label="Batch Sample Size", field=True, value=8, cw=(3,160), minValue=1,
```
Comment (dictoon) [inline scripts/appleseedMaya/renderGlobals.py]: Default should be 16.

File: scripts/appleseedMaya/renderGlobals.py [hunk 4]
```diff
@@ -405,20 +406,24 @@ def create(self):
...
-                                label="Batch Sample Size", numberOfFields=1, enable=adaptiveSampling),
+                            ui=pm.intSliderGrp(
+                                label="Batch Sample Size", field=True, value=8, cw=(3,160), minValue=1, 
+                                fieldMinValue=1, maxValue=128, fieldMaxValue=1000000, enable=adaptiveSampling),
                             attrName="batchSampleSize")
                         self._addControl(
-                            ui=pm.floatFieldGrp(
-                                label="Noise Threshold", numberOfFields=1, enable=adaptiveSampling),
+                            ui=pm.floatSliderGrp(
+                                label="Noise Threshold", field=True, value=0.1, step=0.01, precision=3, cw=(3,160), minValue=0.0001,
```
Comment (dictoon) [inline scripts/appleseedMaya/renderGlobals.py]: Step should be 0.02, I think precision should be 4 (four digits after the decimal point), hard max value should be 10000.

File: scripts/appleseedMaya/renderGlobals.py [hunk 5]
```diff
@@ -427,12 +432,14 @@ def create(self):
...
-                                label="Pixel Filter Size", numberOfFields=1),
+                            ui=pm.floatSliderGrp(
+                                label="Pixel Filter Size", field=True, value=1.5, sliderStep=0.5, precision=1, cw=(3,160), minValue=0.5, 
+                                fieldMinValue=0.5, maxValue=6.0, fieldMaxValue=20.0),
                             attrName="pixelFilterSize")
                         self._addControl(
-                            ui=pm.intFieldGrp(
-                                label="Tile Size", numberOfFields=1),
+                            ui=pm.intSliderGrp(
+                                label="Tile Size", field=True, value=64, cw=(3,160), minValue=8,
```
Comment (dictoon) [inline scripts/appleseedMaya/renderGlobals.py:441]: Hard min tile size should be 1. Hard max tile size should be very large (65536?). It's limited to about 2 billions (2^31-1) in appleseed.studio.

File: scripts/appleseedMaya/renderGlobals.py [hunk 6]
```diff
@@ -427,12 +432,14 @@ def create(self):
                                 enumeratedItem=self._getAttributeMenuItems("pixelFilter")),
                             attrName="pixelFilter")
                         self._addControl(
-                            ui=pm.floatFieldGrp(
-                                label="Pixel Filter Size", numberOfFields=1),
+                            ui=pm.floatSliderGrp(
+                                label="Pixel Filter Size", field=True, value=1.5, sliderStep=0.5, precision=1, cw=(3,160), minValue=0.5, 
+                                fieldMinValue=0.5, maxValue=6.0, fieldMaxValue=20.0),
```
Comment (dictoon) [inline scripts/appleseedMaya/renderGlobals.py]: Soft max value is 4.0 in appleseed.studio.

File: scripts/appleseedMaya/renderGlobals.py [hunk 7]
```diff
@@ -472,32 +479,39 @@ def create(self):
                         limitBounces = mc.getAttr(
                             "appleseedRenderGlobals.limitBounces")
                         self._addControl(
-                            ui=pm.intFieldGrp(
-                                label="Global Bounces", numberOfFields=1, enable=limitBounces),
+                            ui=pm.intSliderGrp(
+                                label="Global Bounces", field=True, value=8, cw=(3,160), minValue=0,
```
Comment (dictoon) [inline scripts/appleseedMaya/renderGlobals.py:483]: I wonder if maxValue should be set to 100 as well? I would tend to think so (for all bounce types).

File: scripts/appleseedMaya/renderGlobals.py [hunk 8]
```diff
@@ -539,20 +554,23 @@ def create(self):
                                 "appleseedRenderGlobals.limitPhotonTracingBounces")
                         self._addControl(
                             ui=pm.intFieldGrp(
-                                label="Max Bounces",  columnWidth=(3, 120), columnAlign = (3, 'left'), 
+                                label="Max Bounces", columnWidth=(3, 120), columnAlign = (3, 'left'),
```
Comment (dictoon) [inline scripts/appleseedMaya/renderGlobals.py]: Remove spaces around `=`?

**Review by dictoon — CHANGES_REQUESTED** [review]
Thanks!!

Partial review only.

File: scripts/appleseedMaya/renderGlobals.py [hunk 9]
```diff
@@ -391,10 +391,11 @@ def create(self):
         with pm.scrollLayout("appleseedScrollLayout", horizontalScrollBarThickness=0):
             with pm.columnLayout("appleseedColumnLayout", adjustableColumn=True, width=columnWidth):
                 with pm.frameLayout(label="Sampling", collapsable=True, collapse=False):
-                    with pm.columnLayout("appleseedColumnLayout", adjustableColumn=True, width=columnWidth):
+                    with pm.columnLayout("appleseedColumnLayout", adjustableColumn=False, width=columnWidth):
                         self._addControl(
-                            ui=pm.intFieldGrp(
-                                label="Render Passes", numberOfFields=1),
+                            ui=pm.intSliderGrp(
+                                label="Render Passes", field=True, value=1, cw=(3,160), minValue=1,
```
[the remark this answers is not in the document]
Comment (Mango-3) [reply scripts/appleseedMaya/renderGlobals.py:397]: Column width of column 3 = 160 units.
Unfortunately, column 1 (the parameter label column) can not be extended this way. Probably it is not modifiable to have a common UI look in all render settings. I checked the Arnold panel and it is the same.  So we are restricted in the length of the parameter names. This is the reason why I have to abbreviate for instance longer labels like `Russian Roulette Start Bounce` with `RR Start Bounce` in the SPPM panel.

File: scripts/appleseedMaya/renderGlobals.py (hunk 1 above)
Comment (Mango-3) [reply scripts/appleseedMaya/renderGlobals.py]: ok

File: scripts/appleseedMaya/renderGlobals.py (hunk 2 above)
Comment (Mango-3) [reply scripts/appleseedMaya/renderGlobals.py:415]: Usually the hard limits are set in the renderglobalsnode.cpp where the parameter are defined.
Here are the input limits for the UI elements (sliders and numeric fields). 
maxValue and minValue are the soft limits for the slider ranges, so a user can see the useful parameter range from the slider range. 
fieldMaxValue and fieldMinValue are the input limits for the numeric fields and override the slider ranges. If a user wants to choose a value outside of the slider range he can just type it then into the numeric field and override the slider (which adapts then to the new input).
If no fieldMaxValue or fieldMinValue are specified here then they are automatically equal to the hard limits of the parameter definition in the c++ code.

File: scripts/appleseedMaya/renderGlobals.py (hunk 3 above)
Comment (Mango-3) [reply scripts/appleseedMaya/renderGlobals.py]: True, this is overridden in the c++ file to 16. I will update it here as well.

File: scripts/appleseedMaya/renderGlobals.py (hunk 4 above)
Comment (Mango-3) [reply scripts/appleseedMaya/renderGlobals.py]: ok

File: scripts/appleseedMaya/renderGlobals.py (hunk 5 above)
Comment (Mango-3) [reply scripts/appleseedMaya/renderGlobals.py:441]: Ok, I will increase max limit.

File: scripts/appleseedMaya/renderGlobals.py (hunk 6 above)
Comment (Mango-3) [reply scripts/appleseedMaya/renderGlobals.py]: ok

File: scripts/appleseedMaya/renderGlobals.py (hunk 7 above)
Comment (Mango-3) [reply scripts/appleseedMaya/renderGlobals.py:483]: The slider has 30 as soft limit but one can override it up to 100 in the numeric field. 
If the slider ranges are large ( goes up to 100) then a small movement of the slider and the bounces change by 5 or 10, which makes it difficult to adjust it for most purposes. 30 bounces is quite large for most scenes.

File: scripts/appleseedMaya/renderGlobals.py (hunk 8 above)
Comment (Mango-3) [reply scripts/appleseedMaya/renderGlobals.py]: yep, true

**Comment by Mango-3** [comment]
Thanks for the detailed review!
I may also add the changes from the SPPM review later as an additional commit to this PR as the SPPM code has been already merged.

File: scripts/appleseedMaya/renderGlobals.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dictoon) [reply scripts/appleseedMaya/renderGlobals.py:397]: Got it, thanks!

**Review by dictoon — APPROVED** [review]

**Pull request merged**